Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift

```diff
@@ -34,163 +34,159 @@ public class StarknetProvider: StarknetProviderProtocol {
         self.init(url: url, urlSession: urlSession)
     }
 
-    private func makeRequest<U>(method: JsonRpcMethod, params: some Encodable = EmptyParams(), receive _: U.Type) async throws -> U where U: Decodable {
-        let rpcPayload = JsonRpcPayload(method: method, params: params)
-
-        var response: JsonRpcResponse<U>
-
-        let config = HttpNetworkProvider.Configuration(url: url, method: "POST", params: [
-            (header: "Content-Type", value: "application/json"),
-            (header: "Accept", value: "application/json"),
-        ])
-
-        do {
-            response = try await networkProvider.send(payload: rpcPayload, config: config, receive: JsonRpcResponse<U>.self)
-        } catch let error as HttpNetworkProviderError {
-            throw error
-        } catch {
-            throw StarknetProviderError.unknownError
+    private func buildRequest<U: Decodable, P: Encodable>(method: JsonRpcMethod, params: P) -> Request<U, P> {
+        let config = HttpNetworkProvider.Configuration(
+            url: url,
+            method: "POST",
+            params: [
+                (header: "Content-Type", value: "application/json"),
+                (header: "Accept", value: "application/json"),
+            ]
+        )
+        return Request<U, P>(method: method, params: params, config: config, networkProvider: networkProvider)
+    }
+
+    /// Batch multiple HTTP requests together into a single HTTP request.
+    ///
+    /// - Parameters
+    ///     - requests: list of HTTP requests to be batched together.
+    ///
+    /// - Returns: batch HTTP request.
+    public func batchRequests<U: Decodable, P: Encodable>(requests: [Request<U, P>]) -> BatchRequest<U, P> {
+        let rpcPayloads = requests.map { request in
+            JsonRpcPayload(method: request.method, params: request.params)
         }
-
-        if let result = response.result {
-            return result
-        } else if let error = response.error {
-            throw StarknetProviderError.jsonRpcError(error.code, error.message, error.data)
-        } else {
-            throw StarknetProviderError.unknownError
+        let config = HttpNetworkProvider.Configuration(
+            url: url,
+            method: "POST",
+            params: [
+                (header: "Content-Type", value: "application/json"),
+                (header: "Accept", value: "application/json"),
+            ]
+        )
+        return BatchRequest<U, P>(rpcPayloads: rpcPayloads, config: config, networkProvider: networkProvider)
+    }
+
+    /// Batch multiple HTTP requests together into a single HTTP request using variadic parameters.
+    ///
+    /// - Parameters
+    ///     - requests: one or more HTTP requests to be batched together.
+    ///
+    /// - Returns: batch HTTP request.
+    public func batchRequests<U: Decodable, P: Encodable>(requests: Request<U, P>...) -> BatchRequest<U, P> {
```

Comment:
When using `batchRequests` with variadic params, deduplication mechanism causes a possibility to receive different numbers of responses than requests. Example:
1. We send a batch request including two calls to get transaction data with the same transaction hashes.
2. We receive an array with one element.

It needs to be handled it some some. Possibly we can even delete the versions with variadic params. Wdhy? @DelevoXDG